### ISSUE TITLE
[lima] K3s split-DNS (again)

### ISSUE THREAD

**Issue by dee-kryvenko** [rancher-sandbox/rancher-desktop#770]
I just gave 0.6.0 a shot and my split-DNS from the host is not working in the container (see https://github.com/rancher-sandbox/rancher-desktop/issues/702).

I've been using Lima 0.7.1 just fine and it worked well (after the https://github.com/lima-vm/lima/pull/281). I doesn't when installed with RD. I did a full reset and manually deleted `lima.yaml` just to make sure.

I can confirm that the DNS resolution works well in the VM but not in the containers. CoreDNS is not able to resolve private addresses. I checked coredns config map and it seems fine (all-defaults). If I do `dig +short @192.168.5.33 private-host` (an IP from `/etc/resolve.conf` in the VM) in the container - it gives nothing, though it works in the VM itself.

**Comment by dee-kryvenko** [comment]
A bit more diagnostic info.

So, my Lima VM `/etc/resolv.conf`:

```
nameserver 192.168.5.3
```

My Lima VM has this line in iptables:

```
-A OUTPUT -d 192.168.5.3/32 -p udp -m udp --dport 53 -j DNAT --to-destination 192.168.5.2:62094
```

With that, in the VM:

```
$ dig @192.168.5.3 +short google.com
172.217.5.206
$ dig @192.168.5.3 +short private.address
a.b.c.d
$ dig @192.168.5.2 -p 62094 +short google.com
172.217.5.206
$ dig @192.168.5.2 -p 62094 +short private.address
a.b.c.d
```

In the container:

```
$ dig @192.168.5.3 +short google.com
172.217.5.206
$ dig @192.168.5.3 +short private.address # nothing
$ dig @192.168.5.2 -p 62094 +short google.com
172.217.5.206
$ dig @192.168.5.2 -p 62094 +short private.address
a.b.c.d
```

So I suspect something's up with the iptables not forwarding `192.168.5.3` IP into the containers.

**Comment by jandubois** [comment]
> I might have figured it out, the `dns.go` implementation seem to only listen for UDP traffic and iptable rule only forwards UDP. Curl and Coredns etc are probably using TCP.

Thank you for digging into this! I thought coredns would prefer udp unless configured otherwise, but obviously we should reply to both tcp and udp for robustness reasons alone.

**Comment by dee-kryvenko** [comment]
Is there any README.md on how to build Lima locally?

**Comment by jandubois** [comment]
> Might be Alpine vs Ubuntu thing

On Ubuntu you are running `systemd-resolved`; you will see that `/etc/resolv.conf` points to `127.0.0.53`, which will respond correctly to both tcp and udp. And `systemd-resolved` will then use udp to query `192.168.5.3`. On Alpine `/etc/resolv.conf` points to `192.168.5.3` directly, where only udp is redirected to the host resolver, and tcp still goes to slirp.

> Is there any README.md on how to build Lima locally?

You need to have `Go` installed. Then all you need is `make all install` inside the repo directory...

**Comment by jandubois** [comment]
> > Is there any README.md on how to build Lima locally?
> 
> You need to have `Go` installed. Then all you need is `make all install` inside the repo directory...

This is for stand-alone lima use only. For putting it into Rancher Desktop the binaries will have to be collected via `hacks/lima-and-qemu.pl` to rewrite the import paths for shared libaries to be relative to the invoking executable.

**Comment by dee-kryvenko** [comment]
Will I not be able to stop RD and then just do `LIMA_HOME=~/Library/Application\ Support/rancher-desktop/lima ./limactl start`?

**Comment by dee-kryvenko** [comment]
So I am thinking something along these lines, but haven't got to how to test it locally yet so it probably doesn't work.

**Comment by dee-kryvenko** [comment]
Ok, it was a typo and now it works in the VM:

```
$ dig +short +notcp @192.168.5.3 private.address
a.b.c.d
a.b.c.e
$ dig +short +tcp @192.168.5.3 private.address
a.b.c.d
a.b.c.e
```

However:

```
$ curl -L https://private.address
curl: (6) Could not resolve host: private.address
```

We are back to square one :(
I ran out of ideas.

**Comment by jandubois** [comment]
> I am now thinking if the host agent should do much more than just the A record proxy. What about DNS delegation, shouldn't it respond to NS requests too?

Yes, it should, but I'm not sure how to do that. That's why I wish the macOS resolver was exposed again via DNS, the way `systemd-resolved` is, so we could just forward to it.

Maybe we can use DNS Service Discovery, but afaik there is only a test tool, and a C API available:

```console
$ dns-sd -Q google.com NS
DATE: ---Mon 11 Oct 2021---
21:23:02.448  ...STARTING...
Timestamp     A/R    Flags if Name                          Type  Class   Rdata
21:23:02.454  Add        3  0 google.com.                   NS     IN     ns3.google.com.
21:23:02.454  Add        3  0 google.com.                   NS     IN     ns4.google.com.
21:23:02.454  Add        3  0 google.com.                   NS     IN     ns1.google.com.
21:23:02.454  Add        2  0 google.com.                   NS     IN     ns2.google.com.
^C
```

A quick Github search shows https://github.com/soh335/go-dnssd for Go bindings, but that repo is 8 years old, with just a single commit 6 years ago. Well, better than nothing.

Please let me know if you are going to attempt to build something along these lines, so I don't have to worry about it! 😄

**Comment by jandubois** [comment]
> > I am now thinking if the host agent should do much more than just the A record proxy. What about DNS delegation, shouldn't it respond to NS requests too?
> 
> Yes, it should, but I'm not sure how to do that.

That being said, I think having a reliable lookup for `A` records that works inside containers, and works for `curl` would give us most of the functionality we need (like 95%). I would think CNAMEs would automatically be resolved by the host resolver. I'm less concerned that `dig` doesn't work correctly inside the VM as it doesn't work correctly on macOS either.

**Comment by AkihiroSuda** [comment]
> I might have figured it out, the `dns.go` implementation seem to only listen for UDP traffic and iptable rule only forwards UDP. Curl and Coredns etc are probably using TCP. I've been able to see with `dig` what `curl` seeing:
> 
> ```
> dig +tcp @192.168.5.3 private.address
> ```
> 
> And the response I am getting must be from QEMU.

Is this confirmed? > `Curl and Coredns etc are probably using TCP.`

TCP DNS isn't widely available, so this seems slightly doubtful

**Comment by jandubois** [comment]
> once we know if the `iptables` fix is good or not

Sorry, I didn't find some contiguous block of quiet time yet to think this through (and do some more research). I hope I'll get around to it tomorrow.

**Comment by dee-kryvenko** [comment]
No worries. It's a KubeCon week after all.

**Comment by dee-kryvenko** [comment]
The above is cool, but it just returns the fastest response (including nxdomain error which is also a response). So it doesn't really work for split-dns...
Found another one https://github.com/greenboxal/dns-heaven - this is purpose built for what we are trying to solve, but it seems unmaintained and also MIT license. Not to mention it is messing up with `/etc/resolve.conf` for whatever reason.
But these two projects gives enough inspiration. We could use `scutils --dns` output and build our own condition forwarder in macos, falling back to the existing `handleDefault` if `scutils` is not available. It's a bit of work but then we will not be in the business of re-implementing DNS server. What do you think?

**Comment by jandubois** [comment]
> I wonder how much more host-agent dns is not behaving up to spec - I don't know the spec that well... I start wondering maybe it shouldn't be a default option?

Compared to the other default option (SLIRP DNS forwarding DNS to just a random nameserver from `/etc/resolv.conf`) I think this is still a better option, especially with your fixes applied.

> Found another one https://github.com/greenboxal/dns-heaven - this is purpose built for what we are trying to solve

Yes, but I'm not convinced of the implementation: it scrapes `scutil --dns` output every few seconds and then still tries to implement the resolver based on those rules.

That's why I was wondering if DNS Service Discovery allows us to hook into the macOS resolver directly, so we would only forward packets. I still think this is worthwhile investigating long-term.

For the near future it might still be best to fix up the host resolver as good as we can.

Personally I find it much more important that containers / apps will resolve addresses correctly than having a full DNS interface that allows `dig` to do all its queries. An app shouldn't care if the `CNAME` was resolved by the local resolver, or one nameserver up the chain; we are essentially just providing a proxy of pre-resolved addresses.

So I think we should concentrate on actual application failures, and not on having all networking tools work properly (they don't work correctly on macOS either).

**Comment by jandubois** [comment]
> > That's why I was wondering if DNS Service Discovery allows us to hook into the macOS resolver directly, so we would only forward packets. I still think this is worthwhile investigating long-term.
> 
> Sorry, I don't know DNS Service Discovery

It is essentially the internal resolver used by macOS itself. It has a [DNSServiceQueryRecord](https://developer.apple.com/documentation/dnssd/1804747-dnsservicequeryrecord) API that can be used to perform DNS queries. It is a C library; I don't know if there is a Go wrapper for it anywhere. Here is some [sample code](https://developer.apple.com/forums/thread/107861) (in C).

You can use the `dns-sd` tool for debugging, but it is not a scrape-able CLI interface for it (if for no other reason than that it keeps listening for updates until you kill it):

```console
$ dns-sd -Q updates.jenkins.io CNAME
DATE: ---Fri 15 Oct 2021---
13:32:41.358  ...STARTING...
Timestamp     A/R    Flags if Name                          Type  Class   Rdata
13:32:41.365  Add        2  0 updates.jenkins.io.           CNAME  IN     mirrors.jenkins.io.
^C
```

So ideally `dns-heaven` would be implemented using this API. I won't be able to do this any time soon, but eventually might look into it if nobody else beats me to it...

> You can't know what Lima/RD might be used for. It becomes more and more popular to pack and run everything as container aka `docker run`. I myself is a maintainer of such a tool https://github.com/plumber-cd/runtainer. What if a network administrator wants to run some DNS tools container to do something?

I'm in full agreement that this is the desirable end-state. But we don't know yet if we can get there at all, or if we do, how long it would take. 

So it should not stand in the way of having something that solves 95% of the users problems much sooner. And I think with your changes, and maybe one more iteration, we are getting close to that. "The perfect is the mortal enemy of the good."

**Comment by jandubois** [comment]
> emulating a DNS that behaves not up to spec

I would like to understand what problems we are facing beyond the fake `AAAA` records.

I have only briefly looked at your openssl cert issue, but am not sure if that is related to DNS at all, unless you are saying that you get an incorrect address for a name (which would be a big problem).

Can you give me some repro instructions so that I can see the difference between the VM (or container) and the host myself? And let me know if you are using Lima from your PR branch, or from `master`.

> I doubt we want to disperse efforts on two far from ideal implementations simultaneously.

Agreed, and I would prefer to avoid polling, if at all possible. I also don't understand how using `scutil --dns` results helps you to avoid the issue you have with the current implementation (probably because I don't understand the issue itself yet).

**Comment by jandubois** [comment]
> I tried in docker-for-desktop cluster - `ubuntu:16.04` behaves the same way (broken). So maybe it's not related to DNS after all.

Yes, I think if it shows the same broken behavior in DD, then this is not solvable by DNS. And as I said before, I don't understand how DNS could make a difference because `openssl` is just connecting to a socket, so unless the IP address is wrong, how could DNS affect the cert?

I suspect the issue is simply the version of `openssl` in the different distros. I think the older version doesn't send a servername with the request, so gets a "random" cert back, and the newer `openssl` will default the `--servername` parameter to the hostname you specified:

```console
$ host updates.jenkins.io
updates.jenkins.io has address 52.202.51.185
updates.jenkins.io is an alias for mirrors.jenkins.io.
updates.jenkins.io is an alias for mirrors.jenkins.io.
$ openssl s_client -connect 52.202.51.185:443 </dev/null 2>/dev/null | grep '^subject='
subject=CN = pkg.origin.jenkins.io
$ openssl s_client -connect updates.jenkins.io:443 </dev/null 2>/dev/null | grep '^subject='
subject=CN = updates.jenkins.io
```

**Comment by jandubois** [comment]
I was wondering if you should also include lookups for `MX` and `SRV` records, given that you have done all the other ones. I don't have any better reason than "for completeness" though.

> Provided you're good with the iptables change.

I'm still trying to understand why the `PREROUTING` rule is necessary, but keep getting lost. But given "the proof of the pudding is in the eating", please mark the PR as ready for review; it is definitely an improvement over what we have right now, so I think it should go in. We can always improve some more later, if necessary.

**Comment by jandubois** [comment]
> misbehaving like the current implementation

Can you spell out explicitly how it is "misbehaving"? Are you just talking about `A` queries automatically resolving `CNAME`s, or is there anything else?

I still feel like we will have a 95%+ solution with what we have right now.

**Comment by jandubois** [comment]
> I updated the PR but I have no idea what these changes actually means and what might be the impact of having both `OUTPUT` and `PREROUTING`. As all the internal k8s routing is based on iptables I am afraid it might break something

Ok, so this [flowchart](https://docs.google.com/drawings/u/1/d/1MtWL8qRTs6PlnJrW4dh8135_S9e2SaawT410bJuoBPk/preview) convinced me that traffic from the cluster (pods) does not get filtered by `nat.OUTPUT` but only by `nat.PREROUTING`, so the same rule needs to be added to both chains, which is exactly what you have done.

I think it fully explains the behaviour you have seen, so I'm fine with merging this part, but still want to review the code that extends the DNS query functionality to other query types.

**Comment by dee-kryvenko** [comment]
> I still feel like we will have a 95%+ solution with what we have right now.

You're right in that - by misbehaving I mean there is likely to be more issues associated with recursive resolution (server claims it can't do recursion for queries other than A/AAAA but in fact it's kind of does it since the system resolver does it) and DNSSEC. `net` lookups does not return the "answer" - just the resulting records. I can't think of what might break exactly (other than DNS diag tools) but I'm certain there will be something. Well, let's see how it goes.

I updated the PR. Sorry I couldn't figure out how do I get the input for `net.LookupSRV` from the query in 10 minutes, so I only added MX for now.

**Comment by dee-kryvenko** [comment]
Ok I believe I've figured SRV out... adding now.
